### PR TITLE
Port FeatureType and S3TypedCSVTarget from Signals.

### PR DESCRIPTION
**Pull request by ncray**
@kkwteh

### CODE REVIEW

File: fireflower/targets.py [hunk 1]
```diff
@@ -133,6 +133,16 @@ def __init__(self, path, compressed=True, kwargs_in=None, kwargs_out=None,
         self.kwargs_out = kwargs_out
         super(S3CSVTarget, self).__init__(path, format)
 
+    @contextmanager
```
Comment (kaloramik) [inline fireflower/targets.py]: Lets not use this, I'm pretty sure it doesnt do anything =( Just inline

File: tests/targets_tests.py [hunk 2]
```diff
@@ -135,6 +137,32 @@ def getenv(v, _):
...
+        types = {'a': FeatureType.int}
+        s = S3TypedCSVTarget(dest_path, types, compressed=True)
+        s.write_csv(df, index=False)
+        read_result = s.read_csv()
+        self.assertDictEqual(df.to_dict(), read_result.to_dict())
 
-
-
+    @mock_s3
+    @tempdir()
```
Comment (kaloramik) [inline tests/targets_tests.py]: dont think tempd is used

File: fireflower/targets.py [hunk 3]
```diff
@@ -195,3 +205,71 @@ def read_csv(self, **kwargs):
                     return pd.read_csv(g, **kwargs)
             else:
                 return pd.read_csv(f, **kwargs)
+
+
+class S3TypedCSVTarget(S3CSVTarget):
+    def __init__(self, path, types, compressed=True,
```
Comment (kaloramik) [inline fireflower/targets.py]: do you wanna just add types to S3CSVTarget? Seems like a lot of reused code.

File: fireflower/targets.py [hunk 4]
```diff
@@ -195,3 +205,71 @@ def read_csv(self, **kwargs):
...
+                df.to_csv(f, **kwargs)
+
+    def read_typed_csv(self, **kwargs):
+        if self.kwargs_in:
+            kwargs = toolz.merge(self.kwargs_in, kwargs)
+        with self._open_reader() as f:
+            dtype = {colname: coltype.serialization_dtype
+                     for colname, coltype in self.types.items()}
+            if self.compressed:
+                df = pd.read_csv(filepath_or_buffer=f,
```
Comment (kaloramik) [inline fireflower/targets.py]: can you read this the same way we read in S3CSVTarget?

 with TextIOWrapper(GzipFile(fileobj=f, mode='rb')) as g:

File: fireflower/targets.py (hunk 1 above)
Comment (ncray) [inline fireflower/targets.py]: done

File: tests/targets_tests.py (hunk 2 above)
Comment (ncray) [inline tests/targets_tests.py]: good catch

File: fireflower/targets.py (hunk 3 above)
Comment (ncray) [inline fireflower/targets.py]: talked offline, added a TODO for this

File: fireflower/targets.py (hunk 4 above)
Comment (ncray) [inline fireflower/targets.py]: thanks, fixed

**Comment by ncray** [comment]
@kaloramik how about now?

**Pull request merged**